### ISSUE TITLE
@Resource field not injected in Jetty Demo

### ISSUE THREAD

**Issue by sbordet** [eclipse/jetty.project#3748]
The Jetty Demo fails to inject `AsyncListenerServlet.MyAsyncListener.maxAmount` with the following exception:

```
javax.naming.NamingException: This context is immutable; remaining name 'com.acme.test.AsyncListenerServlet.MyAsyncListener'
	at org.eclipse.jetty.jndi.NamingContext.createSubcontext(NamingContext.java:371)
	at org.eclipse.jetty.jndi.NamingContext.createSubcontext(NamingContext.java:410)
	at org.eclipse.jetty.jndi.NamingUtil.bind(NamingUtil.java:80)
	at org.eclipse.jetty.plus.jndi.NamingEntry.bindToENC(NamingEntry.java:110)
	at org.eclipse.jetty.plus.jndi.NamingEntryUtil.bindToENC(NamingEntryUtil.java:66)
	at org.eclipse.jetty.annotations.ResourceAnnotationHandler.handleField(ResourceAnnotationHandler.java:158)
	at org.eclipse.jetty.annotations.ResourceAnnotationHandler.doHandle(ResourceAnnotationHandler.java:77)
	at org.eclipse.jetty.annotations.AnnotationIntrospector$AbstractIntrospectableAnnotationHandler.handle(AnnotationIntrospector.java:72)
	at org.eclipse.jetty.annotations.AnnotationIntrospector.introspect(AnnotationIntrospector.java:97)
	at org.eclipse.jetty.annotations.AnnotationDecorator.introspect(AnnotationDecorator.java:61)
	at org.eclipse.jetty.annotations.AnnotationDecorator.decorate(AnnotationDecorator.java:67)
	at org.eclipse.jetty.util.DecoratedObjectFactory.decorate(DecoratedObjectFactory.java:79)
	at org.eclipse.jetty.servlet.ServletContextHandler$Context.createListener(ServletContextHandler.java:1515)
	at org.eclipse.jetty.server.AsyncContextState.createListener(AsyncContextState.java:82)
	at com.acme.test.AsyncListenerServlet.doPost(AsyncListenerServlet.java:98)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
...
```

**Comment by sbordet** [comment]
This is a regression introduced by #3729 and #3739.

**Comment by sbordet** [comment]
@janbartel your input will be great about this.

**Comment by joakime** [comment]
I added more distribution tests to `/test-distribution/` module in branch `jetty-9.4.x-3748-resource-anno-failure`

The following test ...

https://github.com/eclipse/jetty.project/blob/aa9954a9cf97adac52af5ffba7a486cb3b403e79/tests/test-distribution/src/test/java/org/eclipse/jetty/tests/distribution/DemoBaseTests.java#L115-L144

fails with a Http status code response 405 (when expecting a 200)

I need to add into DistributionTests the ability to turn on logging and/or remote debug

**Comment by joakime** [comment]
Whoops, sorry, my mistake in the testcase.
Changing GET to POST now ...

**Comment by joakime** [comment]
I cannot reproduce exception that @sbordet is seeing on my machine.
I'll let jenkins build it on it's multiple JDKs and see if one of them pops up the issue.

**Comment by gregw** [comment]
The jndi patch that Jan and I worked on and that Simone merged did contain a change to the way we lock the java:comp/env context.  Previously only the top level was locked, but reading the spec we realised that it should be a deep lock.   Perhaps we were wrong? or do we need to do the injection earlier, prior to locking?
We are having a look this afternoon

**Comment by gregw** [comment]
A quick fix could be to revert the change to deep locking by removing the parent clause in this change:https://github.com/eclipse/jetty.project/commit/900a036664a2c49ba3e0d8fe726fd3c505b47c8b#diff-f61595e255152e2dde9f37de36cfd1beR1102

A better fix would be to work out how to exempt object decoration from this locking.... difficult as servlets can be created lazily on a request, so long after a lock has been applied.  We can't just unlock temporarily as there could be concurrent requests!

**Comment by joakime** [comment]
Jan opened PR #3754 for this issue

**Comment by janbartel** [comment]
Fixed for 9.4.19.
